check.go

```diff
@@ -66,29 +51,71 @@ func encodeDomainCheck(buf *bytes.Buffer, domains []string, greeting Greeting) e
 		feeURN = ExtFee07
 	}
 
-	if len(feeURN) > 0 {
-		// Extensions
-		buf.WriteString(`<extension>`)
+	supportsLaunch := greeting.SupportsExtension(ExtLaunch) && launchPhase != ""
+	hasExtension := feeURN != "" || supportsLaunch
 
+	if hasExtension {
+		c.buf.WriteString(`<extension>`)
+	}
+
+	if supportsLaunch {
+		c.buf.WriteString(`<launch:check xmlns:launch="` + ExtLaunch + `" type="avail">`)
+		c.buf.WriteString(`<launch:phase>` + launchPhase + `</launch:phase>`)
+		c.buf.WriteString(`</launch:check>`)
+	}
+
+	if len(feeURN) > 0 {
 		// CentralNic fee extension
-		buf.WriteString(`<fee:check xmlns:fee="` + feeURN + `">`)
+		c.buf.WriteString(`<fee:check xmlns:fee="` + feeURN + `">`)
 		for _, domain := range domains {
-			buf.WriteString(`<fee:domain>`)
-			buf.WriteString(`<fee:name>`)
-			xml.EscapeText(buf, []byte(domain))
-			buf.WriteString(`</fee:name>`)
-			buf.WriteString(`<fee:command>create</fee:command>`)
-			buf.WriteString(`</fee:domain>`)
+			c.buf.WriteString(`<fee:domain>`)
+			c.buf.WriteString(`<fee:name>`)
+			xml.EscapeText(&c.buf, []byte(domain))
+			c.buf.WriteString(`</fee:name>`)
+			c.buf.WriteString(`<fee:command>create</fee:command>`)
+			c.buf.WriteString(`</fee:domain>`)
 		}
-		buf.WriteString(`</fee:check>`)
+		c.buf.WriteString(`</fee:check>`)
+	}
 
-		buf.WriteString(`</extension>`)
+	if hasExtension {
+		c.buf.WriteString(`</extension>`)
 	}
 
-	buf.WriteString(xmlCommandSuffix)
+	c.buf.WriteString(xmlCommandSuffix)
 	return nil
 }
 
+func (c *Conn) processDomainCheck(domains []string) (*DomainCheckResponse, error) {
+	err := c.flushDataUnit()
+	if err != nil {
+		return nil, err
+	}
+	var res response_
+	err = c.readResponse(&res)
+	if err != nil {
+		return nil, err
+	}
+
+	// The ARI price extension won't return both availability and price data
+	// in the same response, so we have to make a separate request for price
+	if c.Greeting.SupportsExtension(ExtPrice) {
+		err := encodePriceCheck(&c.buf, domains)
```

Comment:
ARI wrote the price extension. http://ausregistry.github.io/doc/price-1.1/price-1.1.html has all of the details. There is no way to implement the extension without it requiring two messages. We've confirmed this with ARI previously - see #6 for the full discussion.